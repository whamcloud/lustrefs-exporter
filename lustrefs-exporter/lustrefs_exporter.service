[Unit]
Description=Prometheus exporter for Lustre filesystem
Documentation=https://github.com/whamcloud/lustrefs-exporter

[Service]
Environment=RUST_LOG=info
Restart=on-failure
ExecStart=/usr/bin/lustrefs_exporter
MemoryHigh=1750M
MemoryMax=2G

[Install]
WantedBy=multi-user.target
