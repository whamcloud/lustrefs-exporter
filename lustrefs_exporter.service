[Unit]
Description=Prometheus exporter for Lustre filesystem
Documentation=https://github.com/whamcloud/lustrefs-exporter

[Service]
Restart=on-failure
User=prometheus
ExecStart=/usr/bin/lustrefs_exporter

[Install]
WantedBy=multi-user.target

