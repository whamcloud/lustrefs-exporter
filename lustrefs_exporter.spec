Name:           lustrefs_exporter
Version:        0.2.0
Release:        1%{?dist}
Summary:        prometheus exporter for lustre
License:        MIT

Requires(pre): shadow-utils

%description
Prometheus exporter for the Lustre filesystem

%global debug_package %{nil}

%prep

%build
cargo build --release

%install
install -v -d %{buildroot}%{_bindir}
install -v -d %{buildroot}%{_unitdir}
install -v -m 0644 lustrefs_exporter.service %{buildroot}%{_unitdir}
install -v target/release/lustrefs-exporter %{buildroot}%{_bindir}
%{__ln_s} lustrefs-exporter %{buildroot}%{_bindir}/lustrefs_exporter

%files
%{_bindir}/lustrefs-exporter
%{_bindir}/lustrefs_exporter
%{_unitdir}/lustrefs_exporter.service

%pre
getent group prometheus >/dev/null || groupadd -r prometheus
getent passwd prometheus >/dev/null || \
  useradd -r -g prometheus -s /sbin/nologin \
          -c "Prometheus services" prometheus

%post
%systemd_post %{name}.service

%preun
%systemd_preun %{name}.service

%postun
%systemd_postun %{name}.service
