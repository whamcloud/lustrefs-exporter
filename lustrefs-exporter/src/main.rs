// Copyright (c) 2025 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use clap::Parser;
use lustrefs_exporter::{
    Error,
    routes::{
        AppState, app, jobstats_metrics_cmd, lnet_stats_output, lustre_metrics_output,
        net_show_output,
    },
};
use std::{collections::HashMap, net::SocketAddr};

const LUSTREFS_EXPORTER_PORT: &str = "32221";

#[derive(Debug, Parser)]
pub struct CommandOpts {
    /// Port that exporter will listen to
    #[clap(short, long, env = "LUSTREFS_EXPORTER_PORT", default_value = LUSTREFS_EXPORTER_PORT)]
    pub port: u16,

    /// Dump stats as raw string and exit
    #[clap(long, hide = true)]
    dump: bool,
}

/// Dumps Lustre filesystem statistics to stdout
///
/// This function executes several Lustre commands and prints their raw output:
/// - `lctl get_param` with all standard parameters from the parser
/// - `lctl get_param` for jobstats (OST and MDT job statistics)
/// - `lnetctl net show -v 4` for network configuration details
/// - `lnetctl stats show` for network statistics
///
/// # Arguments
/// * `cmd_hdl` - Command handler implementing `RemoteCmd` trait for executing commands
///
/// # Returns
/// * `Ok(())` on successful execution of all commands
/// * `Err(Error)` if any command fails or output cannot be converted to UTF-8
///
/// # Example
/// ```rust
/// use lustrefs_exporter::remote_cmd::LocalCmd;
///
/// dump_stats(&LocalCmd).await?;
/// ```
async fn dump_stats(env_vars: &HashMap<&'static str, String>) -> Result<(), Error> {
    println!("# Dumping lctl get_param output");

    let mut lctl = lustre_metrics_output(env_vars);

    let lctl = lctl.output().await?;

    println!("{}", std::str::from_utf8(&lctl.stdout)?);

    println!("# Dumping lctl get_param jobstats output");

    let mut lctl = jobstats_metrics_cmd(env_vars);

    let lctl = tokio::task::spawn_blocking(move || lctl.output()).await??;

    println!("{}", std::str::from_utf8(&lctl.stdout)?);

    println!("# Dumping lnetctl net show output");

    let mut lnetctl = net_show_output(env_vars);

    let lnetctl = lnetctl.output().await?;

    println!("{}", std::str::from_utf8(&lnetctl.stdout)?);

    println!("# Dumping lnetctl stats show output");

    let mut lnetctl_stats_output = lnet_stats_output(env_vars);

    let lnetctl_stats_output = lnetctl_stats_output.output().await?;

    println!("{}", std::str::from_utf8(&lnetctl_stats_output.stdout)?);

    Ok(())
}

#[tokio::main]
async fn main() -> Result<(), Error> {
    tracing_subscriber::fmt::init();

    let app_state = AppState {
        env_vars: Default::default(),
    };

    let opts = CommandOpts::parse();

    if opts.dump {
        dump_stats(&app_state.env_vars).await?;
    } else {
        let addr = SocketAddr::from(([0, 0, 0, 0], opts.port));

        tracing::info!("Listening on http://{addr}/metrics");

        let listener = tokio::net::TcpListener::bind(("0.0.0.0", opts.port)).await?;

        axum::serve(listener, app(app_state))
            .with_graceful_shutdown(async {
                tokio::signal::ctrl_c().await.ok();
            })
            .await?;
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use lustrefs_exporter::TestEnv;

    use crate::dump_stats;

    #[tokio::test]
    async fn test_dump_stats() {
        let test_env = TestEnv::default();

        dump_stats(&test_env.vars()).await.unwrap();
    }
}
