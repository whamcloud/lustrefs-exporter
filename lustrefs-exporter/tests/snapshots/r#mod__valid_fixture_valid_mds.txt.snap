---
source: lustrefs-exporter/src/main.rs
expression: x
---
# HELP lustre_available_kilobytes Number of kilobytes readily available in the pool.
# TYPE lustre_available_kilobytes gauge
lustre_available_kilobytes{component="mdt",target="ai400x2-MDT0000"} 405704300
lustre_available_kilobytes{component="ost",target="ai400x2-OST0000"} 32526506848
lustre_available_kilobytes{component="ost",target="ai400x2-OST0001"} 32532536272

# HELP lustre_capacity_kilobytes Capacity of the pool in kilobytes.
# TYPE lustre_capacity_kilobytes gauge
lustre_capacity_kilobytes{component="mdt",target="ai400x2-MDT0000"} 414229176
lustre_capacity_kilobytes{component="ost",target="ai400x2-OST0000"} 34750424936
lustre_capacity_kilobytes{component="ost",target="ai400x2-OST0001"} 34750424936

# HELP lustre_connected_clients Number of connected clients.
# TYPE lustre_connected_clients gauge
lustre_connected_clients{component="mdt",target="ai400x2-MDT0000"} 0
lustre_connected_clients{component="mdt",target="ai400x2-MDT0000"} 0

# HELP lustre_exports_dirty_total Total number of exports that have been marked dirty.
# TYPE lustre_exports_dirty_total gauge
lustre_exports_dirty_total{component="ost",target="ai400x2-OST0000"} 0
lustre_exports_dirty_total{component="ost",target="ai400x2-OST0001"} 0

# HELP lustre_exports_granted_total Total number of exports that have been marked granted.
# TYPE lustre_exports_granted_total gauge
lustre_exports_granted_total{component="ost",target="ai400x2-OST0000"} 278208
lustre_exports_granted_total{component="ost",target="ai400x2-OST0001"} 278208

# HELP lustre_exports_pending_total Total number of exports that have been marked pending.
# TYPE lustre_exports_pending_total gauge
lustre_exports_pending_total{component="ost",target="ai400x2-OST0000"} 0
lustre_exports_pending_total{component="ost",target="ai400x2-OST0001"} 0

# HELP lustre_exports_total Total number of times the pool has been exported.
# TYPE lustre_exports_total gauge
lustre_exports_total{component="ost",target="ai400x2-OST0000"} 4
lustre_exports_total{component="ost",target="ai400x2-OST0001"} 4
lustre_exports_total{component="mdt",target="ai400x2-MDT0000"} 15

# HELP lustre_free_kilobytes Number of kilobytes allocated to the pool.
# TYPE lustre_free_kilobytes gauge
lustre_free_kilobytes{component="mdt",target="ai400x2-MDT0000"} 412951760
lustre_free_kilobytes{component="ost",target="ai400x2-OST0000"} 32878173676
lustre_free_kilobytes{component="ost",target="ai400x2-OST0001"} 32884203100

# HELP lustre_health_healthy Indicates whether the Lustre target is healthy or not. 1 is healthy, 0 is unhealthy.
# TYPE lustre_health_healthy gauge
lustre_health_healthy 1

# HELP lustre_inodes_free The number of inodes (objects) available.
# TYPE lustre_inodes_free gauge
lustre_inodes_free{component="mdt",target="ai400x2-MDT0000"} 289693298
lustre_inodes_free{component="ost",target="ai400x2-OST0000"} 274719433
lustre_inodes_free{component="ost",target="ai400x2-OST0001"} 274719491

# HELP lustre_inodes_maximum The maximum number of inodes (objects) the filesystem can hold.
# TYPE lustre_inodes_maximum gauge
lustre_inodes_maximum{component="mdt",target="ai400x2-MDT0000"} 289887952
lustre_inodes_maximum{component="ost",target="ai400x2-OST0000"} 274726912
lustre_inodes_maximum{component="ost",target="ai400x2-OST0001"} 274726912

# HELP lustre_ldlm_cbd_stats Gives information about LDLM Callback service.
# TYPE lustre_ldlm_cbd_stats counter
lustre_ldlm_cbd_stats{operation="req_waittime"} 1
lustre_ldlm_cbd_stats{operation="req_qdepth"} 1
lustre_ldlm_cbd_stats{operation="req_active"} 1
lustre_ldlm_cbd_stats{operation="req_timeout"} 1
lustre_ldlm_cbd_stats{operation="reqbuf_avail"} 3
lustre_ldlm_cbd_stats{operation="ldlm_bl_callback"} 1

# HELP lustre_lnet_mem_used Gives information about Lustre LNet memory usage.
# TYPE lustre_lnet_mem_used gauge
lustre_lnet_mem_used 62041660

# HELP lustre_lock_contended_total Number of contended locks.
# TYPE lustre_lock_contended_total gauge
lustre_lock_contended_total{component="mdt",target="ai400x2-MDT0000"} 32
lustre_lock_contended_total{component="ost",target="ai400x2-OST0000"} 32
lustre_lock_contended_total{component="ost",target="ai400x2-OST0001"} 32

# HELP lustre_lock_contention_seconds_total Time in seconds during which locks were contended.
# TYPE lustre_lock_contention_seconds_total gauge
lustre_lock_contention_seconds_total{component="mdt",target="ai400x2-MDT0000"} 2
lustre_lock_contention_seconds_total{component="ost",target="ai400x2-OST0000"} 2
lustre_lock_contention_seconds_total{component="ost",target="ai400x2-OST0001"} 2

# HELP lustre_lock_count_total Number of locks.
# TYPE lustre_lock_count_total gauge
lustre_lock_count_total{component="mdt",target="ai400x2-MDT0000"} 0
lustre_lock_count_total{component="ost",target="ai400x2-OST0000"} 0
lustre_lock_count_total{component="ost",target="ai400x2-OST0001"} 0

# HELP lustre_lock_timeout_total Number of lock timeouts.
# TYPE lustre_lock_timeout_total counter
lustre_lock_timeout_total{component="mdt",target="ai400x2-MDT0000"} 0
lustre_lock_timeout_total{component="ost",target="ai400x2-OST0000"} 0
lustre_lock_timeout_total{component="ost",target="ai400x2-OST0001"} 0

# HELP lustre_mds_mdt_fld_stats MDS mdt_fld stats.
# TYPE lustre_mds_mdt_fld_stats gauge
lustre_mds_mdt_fld_stats{operation="req_waittime",units="usecs"} 3
lustre_mds_mdt_fld_stats{operation="req_qdepth",units="reqs"} 3
lustre_mds_mdt_fld_stats{operation="req_active",units="reqs"} 3
lustre_mds_mdt_fld_stats{operation="req_timeout",units="secs"} 3
lustre_mds_mdt_fld_stats{operation="reqbuf_avail",units="bufs"} 7
lustre_mds_mdt_fld_stats{operation="fld_query",units="usecs"} 3

# HELP lustre_mds_mdt_stats MDS mdt stats.
# TYPE lustre_mds_mdt_stats gauge
lustre_mds_mdt_stats{operation="req_waittime",units="usecs"} 1502
lustre_mds_mdt_stats{operation="req_qdepth",units="reqs"} 1502
lustre_mds_mdt_stats{operation="req_active",units="reqs"} 1502
lustre_mds_mdt_stats{operation="req_timeout",units="secs"} 1502
lustre_mds_mdt_stats{operation="reqbuf_avail",units="bufs"} 3571
lustre_mds_mdt_stats{operation="mds_connect",units="usecs"} 18
lustre_mds_mdt_stats{operation="obd_ping",units="usecs"} 1484

# HELP lustre_mem_used Gives information about Lustre memory usage.
# TYPE lustre_mem_used gauge
lustre_mem_used 1457821523

# HELP lustre_mem_used_max Gives information about Lustre maximum memory usage.
# TYPE lustre_mem_used_max counter
lustre_mem_used_max 1458113635

# HELP lustre_oss_ost_create_stats OSS ost_create stats.
# TYPE lustre_oss_ost_create_stats gauge
lustre_oss_ost_create_stats{operation="req_waittime",units="usecs"} 4526
lustre_oss_ost_create_stats{operation="req_qdepth",units="reqs"} 4526
lustre_oss_ost_create_stats{operation="req_active",units="reqs"} 4526
lustre_oss_ost_create_stats{operation="req_timeout",units="secs"} 4526
lustre_oss_ost_create_stats{operation="reqbuf_avail",units="bufs"} 9411
lustre_oss_ost_create_stats{operation="ost_statfs",units="usecs"} 4526

# HELP lustre_oss_ost_stats OSS ost stats.
# TYPE lustre_oss_ost_stats gauge
lustre_oss_ost_stats{operation="req_waittime",units="usecs"} 50
lustre_oss_ost_stats{operation="req_qdepth",units="reqs"} 50
lustre_oss_ost_stats{operation="req_active",units="reqs"} 50
lustre_oss_ost_stats{operation="req_timeout",units="secs"} 50
lustre_oss_ost_stats{operation="reqbuf_avail",units="bufs"} 109
lustre_oss_ost_stats{operation="ost_create",units="usecs"} 8
lustre_oss_ost_stats{operation="ost_get_info",units="usecs"} 4
lustre_oss_ost_stats{operation="ost_connect",units="usecs"} 28
lustre_oss_ost_stats{operation="obd_ping",units="usecs"} 10

# HELP lustre_stats_total Number of operations the filesystem has performed.
# TYPE lustre_stats_total counter
lustre_stats_total{component="mdt",operation="statfs",target="ai400x2-MDT0000"} 1691
